File: python/allergies/allergies.py
class Allergies(object):
    def __init__(self, score):
        self.score = score
        self.allergies = {'eggs': 1, 'peanuts': 2, 'shellfish': 4, 
                'strawberries': 8, 'tomatoes': 16, 'chocolate': 32, 
                'pollen': 64, 'cats': 128}

    def is_allergic_to(self, item):
        return (self.score & self.allergies[item]) == self.allergies[item]

    @property
    def lst(self):
        return [k for k, v in self.allergies.items() if self.is_allergic_to(k)]

File: haskell/allergies/src/Allergies.hs
module Allergies (Allergen(..), allergies, isAllergicTo) where

import Data.Bits ((.&.))
import Data.Map.Strict (fromList, findWithDefault, keys)

data Allergen = Eggs
              | Peanuts
              | Shellfish
              | Strawberries
              | Tomatoes
              | Chocolate
              | Pollen
              | Cats
              deriving (Eq, Ord)

m = fromList [(Eggs, 1), (Peanuts, 2), (Shellfish, 4), (Strawberries, 8), (Tomatoes, 16), (Chocolate, 32), (Pollen, 64), (Cats, 128)] 

allergiesMap :: Allergen -> Int
allergiesMap allergen = findWithDefault 0 allergen m

allergies :: Int -> [Allergen]
allergies score = filter (`isAllergicTo` score) (keys m)

isAllergicTo :: Allergen -> Int -> Bool
isAllergicTo allergen score = (.&.) score (allergiesMap allergen) > 0
